smqtk_relevancy/utils/distance_kernel.py

```diff
@@ -0,0 +1,511 @@
+        m = m[np.newaxis]
+
+    log.info("Computing distance kernel")
+    side = m.shape[0]
+    mat = np.ndarray((side, side), dtype=float)
+
+    pr = ProgressReporter(log.debug, 1.0)
+    if row_wise:
+        log.debug("Computing row-wise distances")
+        # For all rows except the last one. We'll have computed all distances by
+        # the time reach m[side-1]
+        if parallel:
+            # noinspection PyShadowingNames
+            def work_func(i):
+                mat[i, i] = dist_func(m[i], m[i])
+                if i < (side - 1):
+                    mat[i + 1:, i] = mat[i, i + 1:] = dist_func(m[i, :],
+                                                                m[i + 1:, :])
+            # Using threading for in-place modification
+            pr.start()
+            for _ in parallel_map(work_func, range(side),
+                                  use_multiprocessing=False):
+                pr.increment_report()
+        else:
+            pr.start()
+            for i in range(side):
+                # Compute col/row wise distances
+                mat[i, i] = dist_func(m[i], m[i])
+                if i < (side-1):
+                    mat[i+1:, i] = mat[i, i+1:] = dist_func(m[i, :], m[i+1:, :])
+                pr.increment_report()
+    else:
+        log.debug("Computing element-wise distances")
+        if parallel:
+            # noinspection PyShadowingNames
+            def work_func(i):
+                mat[i, i] = dist_func(m[i], m[i])
+                # cols to the left of diagonal index for this row
+                for j in range(i):
+                    mat[i, j] = mat[j, i] = dist_func(m[i], m[j])
+            # Using threading for in-place modification
+            pr.start()
+            for _ in parallel_map(work_func, range(side),
+                                  use_multiprocessing=False):
+                pr.increment_report()
+        else:
+            pr.start()
+            for i in range(side):
+                mat[i, i] = dist_func(m[i], m[i])
+                # cols to the left of diagonal index for this row
+                for j in range(i):
+                    mat[i, j] = mat[j, i] = dist_func(m[i], m[j])
+                pr.increment_report()
+    pr.report()
+    return mat
+
+
+def compute_distance_matrix(m1, m2, dist_func, row_wise=False):
+    """
+    Function for computing the pair-wise distance matrix between two arrays of
+    vectors. Both matrices must have the same number of columns.
+    """
+    if m1.ndim == 1:
+        m1 = m1[np.newaxis]
+    if m2.ndim == 1:
+        m2 = m2[np.newaxis]
+    k = np.ndarray((m1.shape[0], m2.shape[0]), dtype=float)
+    if row_wise:
+        # row wise
+        for i in range(m1.shape[0]):
+            k[i, :] = dist_func(m1[i], m2)
+    else:
+        for i in range(m1.shape[0]):
+            for j in range(m2.shape[0]):
+                k[i, j] = dist_func(m1[i], m2[j])
+    return k
+
+
+class DistanceKernel (object):
```

Comment:
This class can likely be removed. I'd say double check that it's not being used, but I don't think it is.